### ISSUE TITLE
WO-021: quality mechanism on

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#26]
> ## NOT ISSUABLE YET - THIS CARD IS A PLACEHOLDER AND ITS CONTENT IS PROVISIONAL
>
> **This work order is issued only AFTER the Phase-2 spec revision.** It is gated on gate **G2**
> and on the lead's spec revision that follows G2. Nothing here may be started before both have
> happened.
>
> **The behaviour of the P2 sketch modules is frozen at that revision, and not before.** PLAN §12.4
> gives WO-021 as a one-line summary; `spec/spec.py` and `gosplan/env/` carry signatures, docstrings
> and constraints for the quality mechanism - an interface, not a decided rule.
>
> **Everything under "Implementation notes" below is PROVISIONAL.** It records what PLAN.md already
> determines and nothing more. The lead replaces it with the revised text at issue time. Do not read
> any of it as settled, and do not present it to anyone else as settled.
>
> **An implementer handed this card in its present form stops and files an AMBIGUITY REPORT**
> (`workorders/AMBIGUITY_TEMPLATE.md`, CONTRACT rule 3) rather than filling the gaps. Inventing the
> undecided half is exactly the failure this card exists to prevent.

> **WO-021** - Phase 2 - Owner **MID-strong** - Difficulty **set by the LEAD at issue** (PLAN §12.4 fixes the tier only) - Gate **G3**
> Card: [`workorders/WO-021.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-021.md)

## What this is

Turns on the quality dimension of production. Today every enterprise's output is a pure quantity;
this card gives each unit a quality level `q`, makes quality cost something to produce, routes it
through the input bundle so a buyer's usable inputs depend on the quality of what it received, and
lets the planner's fulfilment measure see quality only imperfectly (through a measurability
parameter `μ`). It exists because a quantity-keyed plan target is the classic reason quality falls:
if the metric counts units and quality costs effort, the incentive to shed quality is a consequence
of the reward, not a rule anyone wrote. The mechanism is here; the measurement of the resulting
degradation is WO-030's.

## Depends on

- gate G2 - the Phase-1 gate. G2 passes only when the PLAN §4.5 criteria 1-4 hold (DP recovery,
  bunching present/absent, padding elasticity, hygiene) and is signed off by Human + LEAD. This card
  cannot be started before G2 has passed.
- the **P2 spec revision** (LEAD) - the lead's revision of `spec/spec.py` and this card's notes,
  which happens after G2 and which is what freezes the behaviour described below.
- WO-013
- WO-005
- WO-006
- WO-007
- WO-009

Consumed by WO-030 (phenomenon row 3) and WO-031 (the P2 acceptance run).

**Blocked by:** #18 (WO-013), #10 (WO-005), #11 (WO-006), #12 (WO-007), #14 (WO-009), #45 (gate G2)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- `PLAN.md` §§ 2.6, 2.7.3, 2.9.2, 2.9.3, 3, 4.1 (row 3), 4.4 - PLAN.md is git-ignored and supplied
  out of band; it is not in the repository and has no blob link.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rules 4, 5, 6, 7, 9
- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py)
- [`spec/CHANGELOG.md`](https://github.com/Xocas12/gosplan-env/blob/main/spec/CHANGELOG.md)
- [`gosplan/config.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/config.py)
- [`gosplan/params.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/params.py)
- [`gosplan/rng.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/rng.py)
- [`gosplan/env/production.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/production.py)
- [`gosplan/env/planner.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/planner.py)
- [`gosplan/env/reward.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reward.py)
- [`gosplan/env/state.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/state.py)
- [`gosplan/metrics/phenomena.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/metrics/phenomena.py) - the `phenomenon_quality` docstring only
- [`docs/ref_worked_example.md`](https://github.com/Xocas12/gosplan-env/blob/main/docs/ref_worked_example.md)
- **the revised spec text, which does not exist yet - that is what gates this card**

Read nothing else. CONTRACT rule 12.

## Write only

Provisional; the revision fixes the final list.

- `gosplan/env/production.py` - quality routing through the bundle; the `κ_q · q_ik · e_ik` term
- `gosplan/env/planner.py` - `q̄` crediting in delivery; `q̂` into the planner view and into the
  `quality_weighted` branch of `fulfilment_measure`
- `gosplan/env/state.py` - the `quality_acc` period accumulator, if the revision moves it

Not this card: `ref/`, `tests/**`, `spec/spec.py`, `gosplan/metrics/phenomena.py`. If the revision
changes the dynamics at Phase-1 defaults, the **LEAD** regenerates the golden files and writes the
`spec/CHANGELOG.md` entry (CONTRACT rules 1, 2); the implementer never does either.

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

**No new test ids are frozen yet.** The P2 spec revision writes the unit tests and names their ids
in this section; until it does, this field is deliberately empty and the card cannot be issued
(CONTRACT rule 12).

Already binding and already determinate - the frozen Phase-1 suite, which this change must leave
green:

- [ ] **T-U1** per-period per-good conservation to 1e-9 - `tests/unit/test_conservation.py`
- [ ] **T-B7** golden parity with `ref/` - `tests/golden/test_golden_parity.py`
- [ ] **T-B5** welfare blindness - `tests/behavioural/test_welfare_blindness.py`
- [ ] **T-B4** planner blindness - `tests/behavioural/test_planner_blindness.py`
- [ ] **T-B1** no hard-coded pathology - `tests/behavioural/test_no_hardcoded_pathology.py`

Acceptance criterion PLAN already fixes: **at `supply.quality_matters = False` (the Phase-1 default)
every branch this card adds is the exact identity** - `q̄ ≡ 1`, the cost term contributes 0,
`q̂ ≡ 1` - so every golden file passes unchanged. A diff that moves a Phase-1 number is a defect,
not a mechanism.

Completion command, copy-paste:

```bash
uv run ruff check . && uv run ruff format --check . && uv run pytest -q
```

The revision appends the specific `pytest tests/unit/<new test file> -q` invocation. `pytest -q`
honours the `testpaths` of `pyproject.toml` and therefore never collects `tests/acceptance/`
(CONTRACT rule 13).

## Implementation notes

**Provisional until the lead revises the spec.** What follows is only what PLAN.md already
determines; the open questions listed at the end are open, and each of them is an AMBIGUITY REPORT
if it is still open at issue time.

### Formulas PLAN already decides, verbatim

```
cost      c_ik = κ · e_ik² + F · 1[e_ik > 0] + κ_q · q_ik · e_ik            (§2.6)
bundle    X_bj += deliv_bj · q̄_j                                          (§2.7.3)
consumer  consumer_j = Σ_{i∈j} φ_j · shipped_i · q̄_i                      (§2.7.3)
measured  q̂_i = 1 + μ · (q̄_i − 1),  μ = information.quality_measurability  (§2.9.2)
          m_i = R_i · q̂_i   under objective_metric = quality_weighted
logged    val_true_t = Σ_i p_{s(i)} · y_i · q̄_i                           (§2.9.3)
```

Reading of each line:

- **Cost (§2.6).** `κ_q · q_ik · e_ik` is an additional term *inside* the existing per-step cost
  `c_ik`. It is linear in quality and linear in effort, so quality is only expensive when you are
  actually producing. `κ_q = supply.quality_cost`. This is a real production cost, not a reward
  shaping term: CONTRACT rule 4 fixes the production-step reward as exactly `−scale · c_ik`, and
  this change alters `c_ik` itself, which is admitted. Adding a separate quality term anywhere in
  `enterprise_reward` is a rule-4 violation.
- **Bundle (§2.7.3).** A delivery of `deliv_bj` units of good `j` to buyer `b` credits
  `deliv_bj · q̄_j` to the buyer's input stock `X_bj`. Quality is therefore an efficiency units
  adjustment on the *quantity* side: low quality inputs are fewer effective inputs.
- **Consumer sink (§2.7.3).** Final demand receives `Σ_{i∈j} φ_j · shipped_i · q̄_i`, i.e. quality
  scales what reaches the consumer, the same way it scales an intermediate delivery.
- **Measured quality (§2.9.2).** `μ = information.quality_measurability` interpolates between
  quality being invisible to the metric (`μ = 0`, so `q̂ ≡ 1` whatever `q̄` is) and quality being
  fully visible (`μ = 1`, so `q̂ = q̄`). Under `objective_metric = "quality_weighted"` the fulfilment
  measure is `m_i = R_i · q̂_i`; under `objective_metric = "val"` (and the other quantity-keyed
  metrics) it is not, which is precisely the gap that makes degradation profitable.
- **Logging (§2.9.3).** `val_true_t` carries the true `q̄`, never `q̂`. `val_true` and
  `welfare_true` are logged-only quantities: CONTRACT rule 6 and T-B5 forbid them from reaching any
  observation, reward or agent input.

### Sign conventions and units

- `q̄ = 1` is the neutral quality level and the Phase-1 identity. `q̄ > 1` is above-neutral quality,
  `q̄ < 1` is degraded. Every formula above is written so that `q̄ ≡ 1` collapses it to the existing
  Phase-1 expression exactly.
- `q̂` is a *measured* quantity and lives on the planner/metric side only. `q̄` is a *true* quantity
  and lives on the physical side. Crossing them is the bug this mechanism exists to expose, so keep
  them in separate variables with separate names.
- `κ_q` has units of cost per (quality unit × effort unit) and enters `c_ik` additively.

### Edge cases

- `μ = 0` gives `q̂ ≡ 1` regardless of `q̄`. That is the Phase-1 identity and it is also the arm the
  PLAN §4.1 row-3 pipeline check compares against - do not special-case it out of existence, it is a
  used configuration.
- `quality_matters = False` must **short-circuit before any of the above runs**. Not "compute it and
  multiply by one" - an explicit early branch, so the Phase-1 numeric path is bit-for-bit what it
  was and T-B7 golden parity holds with no tolerance argument.
- `q̄ ≡ 1` at `quality_matters = False` means the `κ_q · q · e` term must contribute exactly 0. If
  `κ_q` defaults to 0 in Phase 1 (PLAN §3 gives 0), the term is 0 either way; do not rely on that
  alone, because the short-circuit is what T-B7 actually needs.
- Measurement window: PLAN §4.4, periods `t ≥ 2`. Anything this card accumulates per period must be
  well defined from `t = 0` even though only `t ≥ 2` is measured.

### Tolerances

- T-U1 conservation is to **1e-9** per period per good. A quality-adjusted credit
  (`X_bj += deliv_bj · q̄_j`) changes what "conserved" means: state, in the code and in the
  conservation accounting, whether the quality adjustment is a physical transformation of units or a
  loss term, and make the identity balance either way. If the revision has not said which, that is
  an ambiguity report, not a judgement call.
- T-B7 golden parity is to **1e-9** against `ref/`.

### What PLAN does NOT decide - each is an AMBIGUITY REPORT if still open at issue time

- how `q̄_i` is accumulated from the per-step `q_ik` over the `M` PRODUCE steps (output-weighted
  mean, plain mean, or last value) and how it is reset at the period boundary;
- whether the good-level `q̄_j` credited into a buyer's bundle is a shipment-weighted average over
  the selling sector or something else;
- how quality enters the coverage aggregator, if at all: `X` credited as `deliv · q̄_j` is a
  quantity adjustment, and whether `θ` (the input complementarity of the coverage aggregator) sees
  adjusted or raw stocks changes the economics;
- the Phase-2 value of `quality_cost` `κ_q` - PLAN §3 gives 0 in Phase 1 and no P2 value.

Do not pick any of these. CONTRACT rule 3: choosing "the reasonable default" is a violation.

## Forbidden

- Implementing, computing, plotting or testing any of PLAN §4.1 rows **2, 5, 6, 7**. They are
  **held out** and are computed for the first time in WO-030, in the P2 acceptance run only.
- Implementing `phenomenon_quality` (row 3) - that is WO-030's; only the mechanism is here.
- Any reward term outside CONTRACT rule 4. `κ_q · q · e` is a real cost inside `c_ik`, not a new
  term and not shaping.
- Any function in `gosplan/env/planner.py` taking a `State` other than `make_planner_view`
  (CONTRACT rule 5); any quality quantity reaching an observation other than through the enumerated
  layout of PLAN §2.4 (CONTRACT rule 6).
- Writing a rule that makes quality fall. Degradation under a quantity-keyed objective must be a
  consequence of the reward and the technology (CONTRACT rule 7).
- Direct `numpy.random` / `jax.random` anywhere under `gosplan/env/` (CONTRACT rule 9).
- Changing any PLAN §4.2 locked value, or any daggered Phase-1 value recorded in
  `runs/G1_decision.md`.
- Editing `spec/spec.py`, `tests/**`, `ref/` or regenerating golden files (CONTRACT rules 1, 2).

## Contract rules that bite here

- **Rule 3 - STOP AND REPORT.** Four questions above are undecided. Choosing "the reasonable
  default" for any of them is a violation; file an ambiguity report and end the session.
- **Rule 12 - work orders.** Read only the whitelist, write only the listed files, run the
  completion command verbatim, report in the four-part format.
- **Rule 4 - reward terms.** The quality cost lives inside `c_ik`. No new reward term, no shaping,
  no running reward normalisation.
- **Rule 5 - planner blindness.** `q̂` may reach the planner only through `make_planner_view`. `q̄`
  never reaches it at all.
- **Rule 6 - welfare blindness.** `val_true` carries `q̄` and is logged only; it never appears in an
  observation, a reward or an agent input.
- **Rule 7 - no hard-coded pathology.** No transition rule may make quality fall.
- **Rule 9 - RNG.** All randomness through `rng.draw(seed_env, purpose, *indices)`.
- **Rules 1 and 2 - frozen spec and frozen tests.** `spec/spec.py` and `tests/**` are not yours; a
  test that looks wrong gets an ambiguity report.

## Definition of done

- [ ] Every must-pass test green, including the frozen Phase-1 suite unchanged
- [ ] At `quality_matters = False` every golden file passes unchanged (no Phase-1 number moved)
- [ ] `uv run ruff check . && uv run ruff format --check .` clean
- [ ] `uv run python scripts/contract_guard.py --all` clean
- [ ] No file outside "Write only" created, edited or deleted
- [ ] CI `ci-ok` green on the PR
- [ ] The four-part report posted as the PR body: 1) tests passed/failed with output, 2) ambiguity
      reports filed, 3) files changed, 4) anything you were tempted to invent and did not
